### DISCUSSION TITLE
Using Git LFS for mobile applications

### DISCUSSION THREAD

**Discussion by shahzadmajeed** [MobileNativeFoundation/discussions#192] [General]
Hi, 

There wasn't any dedicated discussion thread for Git LFS topic so I thought I should create one. I am looking to setup Git LFS for our monorepo where uses a mix of swift packages (local & remote) and cocoapods as dependencies. Almost all of our cocoapods are pre-built (via. a custom script written on top of https://github.com/grab/cocoapods-binary-cache) and cached in same repository. This helps with reducing build times etc.

But binary frameworks (other binary) files actually increase repo size much faster when we update them. So, I'm planning to add Git LFS support. 

I did this on an experimental clone of the repo and it's introducing different sort of random issues i.e., sometime not being able to track all files in .framework & .xcframework. Sometime it doesn't recognize binary file at all. I have seen issues with push/pull as well.

So, I'm wondering if somebody here have used Git LFS and what is your experience with it? How do you manage .gitattributes?

Here is my example .gitattributes file

```
**/*.zip filter=lfs diff=lfs merge=lfs -text
**/*.a filter=lfs diff=lfs merge=lfs -text
**/*.xcframework/**/* filter=lfs diff=lfs merge=lfs -text
**/*.framework/**/* filter=lfs diff=lfs merge=lfs -text
**/*.mp4 filter=lfs diff=lfs merge=lfs -text
**/*.mp3 filter=lfs diff=lfs merge=lfs -text
```

**Comment by tcamin** [comment]
At Subito, we have been utilizing GitHub's LFS servers on our enterprise instance for several years now, and have never encountered any significant issues with the setup using a .gitattributes file very similar to the one you posted. 

Although this may not directly address your problem, I'd like to share more about our setup as it could provide some valuable insights.

To streamline prebuilding, we developed [PodBuilder](https://github.com/Subito-it/PodBuilder), which besides prebuilding comes with additional features that allows to quickly switch between prebuilt and non-prebuilt versions of a dependency. The tool creates a local podspec for each prebuilt dependency, which is an exact replica of the original podspec, but with the prebuilt framework vendored instead of the source code files. In a simplified scenario where the dependency is also included in your monorepo, switching from the prebuilt to non-prebuilt version becomes just a matter of changing the pod's :path => in the podspec. 

A dedicated script will take care to update the app Podfile switching to the prebuilt version only for dependencies that did not change since they were last prebuilt. As a result, every revision of the main branch contains a functioning version of the project, with only a few dependencies potentially not yet rebuilt by our CI, which is responsible to generate xcframeworks and push them to LFS.

Depending on your team size I believe that the prebuilt + LFS approach in general is a viable solution with several advantages and quick wins:
- It's simple to get started and allows to get immediate benefits by starting to prebuild only dependencies that tend to change less frequently (e.g Firebase)
- The dev experience doesn't change since git LFS hides everything behind a single `git pull` command that requires little setup particularly if you use GitHub's LFS instance
- While lfs space doesn't come for free you trade disk space with developer time
- Prebuilding provides a tangible and measurable reason to encourage modularization, as build times decrease significantly. At least for us it has been a great incentive to get started with the modularization of the app
- When modularizing it is common practice to create a separate project for each dependency, which allows you to work on a smaller part of the application in isolation. If a module depends upon another you can reuse the same prebuilt xcframeworks used in the main app
- This way of prebuilding results in constant main app build times as long as new features get added to separate modules

Of course there are downsides
- Files you push to LFS cannot be deleted, at least not with potentially undesidered effects
- The overall size can increase really rapidly particularly if you start modularizing in a way where caches need to be invalidated very frequently
- You generally prebuild dependencies with library evolution support to be able to use them on any version of Xcode. This however makes you bump into some .swiftinterface shenanigans since xcodebuild doesn't always generate them as expected
- Someone, one day, will come and knock at your Slack asking why on earth you're using so much space

To sum up, while I'm happy with the current setup we also dedicated lots of effort to minimize the downsides mentioned above, so it might not be always the best choice. For a quick win definitely, for a more long term solution one might want to evaluate other alternatives that offer integrated remote caching solutions, like Tuist or Bazel (which afaik do not use lfs).
